src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+//! while the former provide views for subgraphs within a single level of the
+//! hierarchy.
+//!
+
+// TODO:
+// //! This module exposes the [`SiblingView`] trait, which is currently
+// //! implemented by the [`SiblingSubgraph`] struct in this module, as well as
+// //! the [`SiblingGraph`] hierarchical view.
+
+use itertools::{Either, Itertools};
+use portgraph::{algorithms::ConvexChecker, view::Subgraph, PortView};
+use thiserror::Error;
+
+use crate::{
+    ops::{OpTag, OpTrait},
+    types::{AbstractSignature, EdgeKind},
+    Direction, Hugr, Node, Port, SimpleReplacement,
+};
+
+use super::{sealed::HugrInternals, HugrView};
+
+/// A non-empty convex subgraph of a HUGR sibling graph.
+///
+/// A HUGR region in which all nodes share the same parent. Unlike
+/// [`super::SiblingGraph`],  not all nodes of the sibling graph must be
+/// included. A convex subgraph is always an induced subgraph, i.e. it is defined
+/// by a set of nodes and all edges between them.
+
+/// The incoming boundary (resp. outgoing boundary) is given by the input (resp.
+/// output) ports of the subgraph that are linked to nodes outside of the subgraph.
+/// The signature of the sugbraph is then given by the types of the incoming
+/// and outgoing boundary edges. Given a replacement with the same signature,
+/// a [`SimpleReplacement`] can be constructed to rewrite the subgraph with the
+/// replacement.
+///
+/// The ordering of the nodes in the subgraph is irrelevant to define the convex
+/// subgraph, but it determines the ordering of the boundary signature.
+#[derive(Clone, Debug)]
+pub struct SiblingSubgraph<'g, Base> {
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a HUGR.
+    ///
+    /// The subgraph is given by the sibling graph of the root. If you wish to
+    /// create a subgraph from another root, wrap the argument `region` in a
+    /// [`super::SiblingGraph`].
+    ///
+    /// This panics if the sibling graph is empty.
+    pub fn from_sibling_graph(sibling_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let root = sibling_graph.root();
+        let nodes = sibling_graph.children(root).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty graph");
+        }
+        Self {
+            base: sibling_graph,
+            nodes,
+        }
+    }
+
+    /// A sibling subgraph from a [`crate::ops::OpTag::DataflowParent`]-rooted HUGR.
+    ///
+    /// The subgraph is given by the nodes between the input and output
+    /// children nodes of the parent node. If you wish to create a subgraph
+    /// from another root, wrap the `region` argument in a [`super::SiblingGraph`].
+    ///
+    /// This panics if the DFG graph is empty.
+    pub fn from_dataflow_graph(dfg_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
```

Comment:
You can restrict the root node to dfgs here.
```suggestion
        Base: HugrView<RootHandle = DfgID>,
```
You may need to edit the siblinggraph defs in your tests;
`SiblingGraph<'_>` → `SiblingGraph<'_, _>`